File: object-oriented-programming/virtual-world-python/plants/sosnowskys_hogweed.py
from plants.plant import Plant
from species import Species
import random

class Sosnowskys_Hogweed(Plant) :

    def __init__(self, world, x = None, y = None) :
        if x == None and y == None :
            super().__init__(Species.SOSNOWSKYS_HOGWEED, 0, 0, 10, world, 1)
            width = world.width
            height = world.height

            while(True):
                new_x = random.randint(0, width - 1)
                new_y = random.randint(0, height - 1)
                if self._world.is_free(new_x, new_y) :
                    break
                else :
                    continue

            self._world.set_organism(None, self.x, self.y)
            self._x = new_x
            self._y = new_y
            self._world.set_organism(self, self.x, self.y)
            return
        
        super().__init__(Species.SOSNOWSKYS_HOGWEED, x, y, 10, world, 1)

    def action(self) :
        dx = [0, 1, 1, 1, 0, -1, -1, -1]
        dy = [1, 1, 0, -1, -1, -1, 0, 1]
        dx2 = [0, 1, 0, -1]
        dy2 = [1, 0, -1, 0]
        for i in range(8) :
            newX = self.x + dx[i]
            newY = self.y + dy[i]
            if self._world.in_bounds(newX, newY) == False : continue 
            neighbour = self._world.get_organism(newX, newY)
            if neighbour != None and neighbour.species != Species.SOSNOWSKYS_HOGWEED and neighbour.species != Species.BELLADONNA and neighbour.species != Species.GUARANA and neighbour.species != Species.SOW_THISTLE and neighbour.species != Species.GRASS and neighbour.species != Species.CYBERSHEEP :
                self._world.report(f"Sosnowsky's Hogweed killed {neighbour.species.display_name()} !")
                neighbour.mark_dead()
                self._world.set_organism(None, newX, newY)
        
        for i in range(self._sow_attempts) : 
            direction = random.randint(0, 3)
            newX = self.x + dx2[direction]
            newY = self.y + dy2[direction]
            if self._world.in_bounds(newX, newY) and self._world.get_organism(newX, newY) == None and random.randint(0, 2) == 0 :
                self.sow(newX, newY)
                return

    def collision(self, other):
        return super().collision(other)

    def draw_color(self):
        return self.species.color
    
    def sow(self, x, y):
        return Sosnowskys_Hogweed(self._world, x, y)

File: object-oriented-programming/virtual-world-java/src/Game/Model/Plants/SosnowskysHogweed.java
package Game.Model.Plants;

import Game.Model.Organism;
import Game.Model.Species;
import Game.Model.World;

import java.awt.*;
import java.util.Random;

public class SosnowskysHogweed extends Plant {
    Random random = new Random();

    public SosnowskysHogweed(int x, int y, World world){
        super(Species.SOSNOWSKYS_HOGWEED, x, y, 10, world, 1);
    }

    public SosnowskysHogweed(World world){
        super(Species.SOSNOWSKYS_HOGWEED, 0, 0, 10, world, 1);
        int width = world.getWidth();
        int height = world.getHeight();

        int newX, newY;

        do {
            newX = random.nextInt(width);
            newY = random.nextInt(height);
        } while(!world.isFree(newY, newX));

        world.setOrganism(null, y, x);
        x = newX;
        y = newY;
        world.setOrganism(this, y, x);
    }

    public void action(){
        int[] dx = {0, 1, 1, 1, 0, -1, -1, -1};
        int[] dy = {1, 1, 0, -1, -1, -1, 0, 1};
        int newX, newY;
        for(int i = 0; i < 8; i++){
            newX = getX() + dx[i];
            newY = getY() + dy[i];
            if(!world.inBounds(newY, newX)) continue;
            Organism neighbour = world.getOrganism(newY, newX);
            if(neighbour != null && (neighbour.getSpecies() != Species.SOSNOWSKYS_HOGWEED &&
                    neighbour.getSpecies() != Species.BELLADONNA &&
                            neighbour.getSpecies() != Species.GUARANA &&
                                    neighbour.getSpecies() != Species.SOW_THISTLE &&
                                            neighbour.getSpecies() != Species.GRASS)){
                world.report("Sosnowsky's Hogweed killed " + neighbour.getSpecies().displayName() + "!");
                neighbour.markDead();
                world.setOrganism(null, newY, newX);
            }
        }

        for(int i = 0; i < sowAttempts; i++){
            int direction = random.nextInt(4);
            newX = x + dx[direction];
            newY = y + dy[direction];
            if(world.inBounds(newY, newX) && world.getOrganism(newY, newX) == null && random.nextInt(2) == 0){
                sow(newY, newX);
                return;
            }
        }
    }

    public final Color draw(){
        return species.displayColor();
    }

    public SosnowskysHogweed sow(int y, int x){
        return new SosnowskysHogweed(x, y, world);
    }
}
